### PR TITLE
add signal ports to gas canisters

### PR DESCRIPTION
**Pull request by Ilya246**
## About the PR

adds toggle, open and close input signal ports to gas canisters, also adds a state output port

## Why / Balance

allows for signal-controlled deployment of gas via canister release and not only signal valve

## Technical details

moves admin logging of canister valve alterations to a new method, and also handles sending state signals via that method
when receiving a network signal and the state it would set the canister to is not identical to the current state, change the valve via said method

## Media

https://github.com/space-wizards/space-station-14/assets/57039557/a17c526e-3234-488b-abbb-fa5061419c29

- [x] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

**Changelog**

:cl:
- add: Gas canisters now have signal output and input ports.

### CODE REVIEW

**Comment by Anzarot121** [comment]
Atmos bomb has never been so user friendly.

**Comment by Ilya246** [comment]
if this is "bad" due to maxcaps, just nerf maxcaps

**Comment by Ilya246** [comment]
> Making maxcaps not a suicide bomb. Also gassing the bar w/n2o would be pretty common place.

you can already remotely detonate maxcaps, and i don't really see how this makes it easier to gasflood

**Comment by Anzarot121** [comment]
> if this is "bad" due to maxcaps, just nerf maxcaps
It's not bad. On the server where I play, only one person can quickly make them.

**Comment by deltanedas** [comment]
you can already gas the bar without a canister at all just use vent and pump

**Comment by PJB3005** [comment]
My suggestion: make it so that you need to add some jury-rigged signal receiver to the canister to make it able to be remotely controlled. This would make it blatantly obvious on the sprite that a canister may be remotely rigged.

**Comment by Ilya246** [comment]
> My suggestion: make it so that you need to add some jury-rigged signal receiver to the canister to make it able to be remotely controlled. This would make it blatantly obvious on the sprite that a canister may be remotely rigged.

there isn't really any need, canisters show on their sprite when there's something inside, and anyone who would get suspicious of a remotely rigged canister would also get suspicious of a canister with a tank in it, especially knowing it might be rigged

**Comment by PJB3005** [comment]
> > My suggestion: make it so that you need to add some jury-rigged signal receiver to the canister to make it able to be remotely controlled. This would make it blatantly obvious on the sprite that a canister may be remotely rigged.
> 
> there isn't really any need, canisters show on their sprite when there's something inside, and anyone who would get suspicious of a remotely rigged canister would also get suspicious of a canister with a tank in it, especially knowing it might be rigged

Tanks can be rigged to do something remotely while inside a canister? What? 

The point is that a canister on its own is an inert object and at least needs somebody to touch it, regardless of the contents. This PR changes that.

**Comment by hord-brayden** [comment]
> Air alarms make no sense, canisters are a mobile object, air alarms are for static configuration.
> 
> I genuinely do not see the need for advanced remote control of canisters. They're something you use to MOVE gas, they aren't an active piece of machinery themselves (like the pipe network you wrench them into). 

I disagree. With reference to the storage tanks they're  storage devices which serve the purpose of storing mixes or a subset, much like pipes operate currently. 

Additionally if filling multiple tanks with a mixture, the only way to verify the contents is to run around with an analyzer. The same goes for pipes. It would be very useful when producing a product to have a central location to see the output rather than scurrying about from pipe to pipe, can to can, ensuring everything is fine - as there's no way to determine a pipes contents beyond manually checking it emptying it into a burn room or similar analogue. 

Just my thoughts

**Review by juliangiebel — CHANGES_REQUESTED** [review]
Implementation looks good, there is a minor code style issue though.
I like this feature but I'm not sure about the implications of remotly opening gas canisters

**Comment by juliangiebel** [comment]
Lmao there was a whole conversation going on while I was reviewing >.<

**Comment by PJB3005** [comment]
> > Air alarms make no sense, canisters are a mobile object, air alarms are for static configuration.
> > 
> > I genuinely do not see the need for advanced remote control of canisters. They're something you use to MOVE gas, they aren't an active piece of machinery themselves (like the pipe network you wrench them into). 
> 
> I disagree. With reference to the storage tanks they're  storage devices which serve the purpose of storing mixes or a subset, much like pipes operate currently. 
> 
> Additionally if filling multiple tanks with a mixture, the only way to verify the contents is to run around with an analyzer. The same goes for pipes. It would be very useful when producing a product to have a central location to see the output rather than scurrying about from pipe to pipe, can to can, ensuring everything is fine - as there's no way to determine a pipes contents beyond manually checking it, emptying it into a burn room, or a similar analogue. 
> 
> Just my thoughts 

That is what pipe meters are for, vouch for those instead.

**Comment by Ilya246** [comment]
> > > My suggestion: make it so that you need to add some jury-rigged signal receiver to the canister to make it able to be remotely controlled. This would make it blatantly obvious on the sprite that a canister may be remotely rigged.
> > 
> > 
> > there isn't really any need, canisters show on their sprite when there's something inside, and anyone who would get suspicious of a remotely rigged canister would also get suspicious of a canister with a tank in it, especially knowing it might be rigged
> 
> Tanks can be rigged to do something remotely while inside a canister? What?
> 
> The point is that a canister on its own is an inert object and at least needs somebody to touch it, regardless of the contents. This PR changes that.

i was referring to atmos bombs, they need a canister to be opened while inside to detonate
as for that, to me canisters don't really seem like simple gas containers, they appear to contain at least enough electronics to handle access and are rather expensive (1000) for whatever reason

**Comment by PJB3005** [comment]
I mean if you ask me, canisters having access locks is also pretty stupid. But I'm an SS13 boomer so.

**Comment by juliangiebel** [comment]
I agree with PJBs suggestions. This should be a rigged device you put on the cannister.
This may even be inplemented as a general feature of attaching a device to things that have a slot for it (so not like c4)

File: Content.Server/Atmos/Piping/Unary/EntitySystems/GasCanisterSystem.cs [hunk 1]
```diff
@@ -152,13 +159,44 @@ private void OnCanisterChangeReleaseValve(EntityUid uid, GasCanisterComponent ca
         {
             containedGasDict.Add((Gas)i, canister.Air.Moles[i]);
         }
+        
+        if (player != null)
+            _adminLogger.Add(LogType.CanisterValve, impact, $"{ToPrettyString(player.Value):player} set the valve on {ToPrettyString(uid):canister} to {state:valveState} while it contained [{string.Join(", ", containedGasDict)}]");
+        else
+            _adminLogger.Add(LogType.CanisterValve, impact, $"Valve on {ToPrettyString(uid):canister} has been set to {state:valveState} while it contained [{string.Join(", ", containedGasDict)}]");
```
[the remark this answers is not in the document]
Comment (Ilya246) [reply Content.Server/Atmos/Piping/Unary/EntitySystems/GasCanisterSystem.cs]: done

**Comment by PJB3005** [comment]
You do understand the distinction between an AI being able to open any canister it can see, vs a borg having to move up to the canister like a human, yes?

**Comment by hord-brayden** [comment]
I think you misunderstand, with this PR, they'd link it with a multitool and go hook it up to something  - in effect a door. When open opens/closes; open can.

I think the distinction between when someone is trying to argue or be helpful is currently unrecognizable. My only intent is the latter, apologies if you've gotten a different impression

**Comment by deltanedas** [comment]
borgs have aa so they can just unlock no need to link it

**Comment by PJB3005** [comment]
Did we not give borgs the ability to remotely control things?

**Comment by hord-brayden** [comment]
Sorry if this derailed this PR. I'm just salty from prior maxcappery, and that's all I see here in my future! As always just my $0.02

In my mind, giving them a multi tool and the ability to build/link to new or existing structures gives them remote access to trigger this. It's a hacky way, but I'm unfamiliar with Borgs and if there's a remote signaller board or not?

**Comment by felineace** [comment]
I can personally see this being used to help simplify and compact air refilling setups on shuttles, which, since we are moving away from small fans for some reason, i can see as particularly good qol for those that actually want to build and encourage construction of shuttles

**Comment by TemporalOroboros** [comment]
@Partmedia 
Can you post what changes did you want made? I don't think they can make those changes if they don't know what they are.

**Comment by Ilya246** [comment]
> @Partmedia Can you post what changes did you want made? I don't think they can make those changes if they don't know what they are.

it was discussed on discord, i'll possibly work on it later

**Comment by Partmedia** [comment]
Please re-open if you plan to continue working on this.

**Comment by ZombieChibiXD** [comment]
Sorry for necro but is this being worked on or dropped?

**Comment by Ilya246** [comment]
> Sorry for necro but is this being worked on or dropped?

if someone makes a sprite for it i might maybe revive it

**Pull request closed without merging**